tokio/src/io/util/mem.rs

```diff
@@ -0,0 +1,178 @@
+//! IO over in-process memory
+
+use crate::io::{AsyncRead, AsyncWrite};
+
+use std::{task::{self, Poll, Waker}, pin::Pin, sync::{Arc, Mutex}};
+use bytes::{Buf, BytesMut};
+
+/// A bidirectional pipe to read and write bytes in memory.
+///
+/// A pair of `MemStream`s are created together, and they act as a "channel"
+/// that can be used as in-memory IO types. Writing to one of the pairs will
+/// allow that data to be read from the other, and vice versa.
+///
+/// # Example
+///
+/// ```
+/// # async fn ex() -> std::io::Result<()> {
+/// # use tokio::io::{AsyncReadExt, AsyncWriteExt};
+/// use tokio::io::MemStream;
+///
+/// let (mut client, mut server) = MemStream::pair();
+///
+/// client.write_all(b"ping").await?;
+///
+/// let mut buf = [0u8; 4];
+/// server.read_exact(&mut buf).await?;
+/// assert_eq!(&buf, b"ping");
+///
+/// server.write_all(b"pong").await?;
+///
+/// client.read_exact(&mut buf).await?;
+/// assert_eq!(&buf, b"pong");
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct MemStream {
```

Comment:
I actually like `Pipe` also for the public name - or `InMemoryPipe`.

Reply:
I also liked `Pipe` (and `pipe()`), but had these concerns:

- I didn't want people thinking this is backed by a Unix pipe.
- A Unix pipe is unidirectional, and these types are bidirectional.